Subject: [PATCH] Limit the number of start or bind service

Symptom:
system_server Watchdog Timeout.

Root Cause:
Bad behavior of app keeps bind/start Service without
unbind/stop it. AMS needs to clean the records once process
died and watchdog will monitor certain threads if they block
for a long time. If the leaked count is too large, AMS needs
to take more time to handle it, then trigger framework
restart to prevent device hung.

Solution:
Limit the number of start or bind service.

Test:
Write a app to bind or start 10001 services without unbind
or stop it, the 10001 service cannot be binded or started
successfully.

diff --git a/services/core/java/com/android/server/am/ActiveServices.java b/services/core/java/com/android/server/am/ActiveServices.java
index ae1aef6..075d397 100755
--- a/services/core/java/com/android/server/am/ActiveServices.java
+++ b/services/core/java/com/android/server/am/ActiveServices.java
@@ -119,6 +119,8 @@
     // allowing the next pending start to run.
     static final int BG_START_TIMEOUT = 15*1000;
 
+    static final int MAX_SERVICE_COUNT = 10000;
+
     final ActivityManagerService mAm;
 
     // Maximum number of services that we allow to start in the background
@@ -379,6 +381,12 @@
             }
         }
 
+        if (r.deliveredStarts.size() + r.pendingStarts.size() >= MAX_SERVICE_COUNT) {
+            Slog.e(TAG_SERVICE, r + " has already start " + MAX_SERVICE_COUNT
+                    + " services. Not allow to start more.");
+            return null;
+        }
+
         if (unscheduleServiceRestartLocked(r, callingUid, false)) {
             if (DEBUG_SERVICE) Slog.v(TAG_SERVICE, "START SERVICE WHILE RESTART PENDING: " + r);
         }
@@ -989,6 +997,14 @@
 
         final long origId = Binder.clearCallingIdentity();
 
+        IBinder ibinder = connection.asBinder();
+        ArrayList<ConnectionRecord> list = s.connections.get(ibinder);
+        if (list != null && list.size() >= MAX_SERVICE_COUNT) {
+            Slog.e(TAG_SERVICE, s + " has already bind " + MAX_SERVICE_COUNT
+                    + " connections. Not allow to bind more.");
+            return 0;
+        }
+
         try {
             if (unscheduleServiceRestartLocked(s, callerApp.info.uid, false)) {
                 if (DEBUG_SERVICE) Slog.v(TAG_SERVICE, "BIND SERVICE WHILE RESTART PENDING: "
